### ISSUE TITLE
Disabled MenuItem's onClick is still triggered

### ISSUE THREAD

**Issue by trend-derek-hawker** [cheton/styled-ui#613]
The `onClick` event for disabled MenuItems will trigger, which is unexpected because the same wouldn't happen on disabled buttons.

**How to reproduce**:
Add `disabled` and `onClick` to MenuItem component

**Expectation**:
`onClick` should not trigger for disabled MenuItem

**Actual**:
`onClick` is triggered

**Video**:

https://user-images.githubusercontent.com/79874774/183098334-9d2f6d1a-ff9d-4c45-9aff-016e41b1304e.mov

**Code to Reproduce**
```
<Menu>
  <MenuButton>Select an option</MenuButton>
  <MenuList>
    <MenuItem>List item</MenuItem>
    <MenuItem disabled onClick={()=>alert("oh no")}>List item</MenuItem>
    <MenuItem>List item</MenuItem>
  </MenuList>
</Menu>
```

**Comment by cheton** [comment]
This can be fixed internally by checking the `disabled` prop to determine whether it should call the `onClick` handler.

For example:

```jsx
onClick={callEventHandlers(
  disabled ? event => event.preventDefault() : onClick,
  handleClick
)}
```

Unfortunately, there're 30+ places that use `callEventHandlers` and they might have to behave the same for the disabled state. 

Please allow me to take some time to figure out what needs to change.

```
./menu/SubmenuList.js:  const onMouseEnter = callEventHandlers(onMouseEnterProp, () => {
./menu/SubmenuList.js:  const onMouseLeave = callEventHandlers(onMouseLeaveProp, () => {
./menu/SubmenuToggle.js:  const handleMouseEnter = callEventHandlers(onMouseEnterProp, (event) => {
./menu/SubmenuToggle.js:  const handleMouseLeave = callEventHandlers(onMouseLeaveProp, (event) => {
./menu/MenuToggle.js:  const handleClick = callEventHandlers(onClickProp, (event) => {
./menu/MenuToggle.js:  const handleKeyDown = callEventHandlers(onKeyDownProp, event => {
./menu/MenuItem.js:      onClick={callEventHandlers(onClick, event => {
./menu/MenuItem.js:      onKeyDown={callEventHandlers(onKeyDown, event => {
./menu/MenuContent.js:    onBlur: callEventHandlers(onBlurProp, handleBlur),
./menu/MenuContent.js:    onKeyDown: callEventHandlers(onKeyDownProp, handleKeyDown),
./tabs/Tab.js:  const handleClick = callEventHandlers(onClick, (event) => {
./popover/PopoverTrigger.js:          onClick: callEventHandlers(ownProps?.onClick, clickTriggerHandler.onClick),
./popover/PopoverTrigger.js:          onKeyDown: callEventHandlers(ownProps?.onKeyDown, clickTriggerHandler.onKeyDown),
./popover/PopoverTrigger.js:          onBlur: callEventHandlers(ownProps?.onBlur, hoverTriggerHandler.onBlur),
./popover/PopoverTrigger.js:          onFocus: callEventHandlers(ownProps?.onFocus, hoverTriggerHandler.onFocus),
./popover/PopoverTrigger.js:          onKeyDown: callEventHandlers(ownProps?.onKeyDown, hoverTriggerHandler.onKeyDown),
./popover/PopoverTrigger.js:          onMouseEnter: callEventHandlers(ownProps?.onMouseEnter, hoverTriggerHandler.onMouseEnter),
./popover/PopoverTrigger.js:          onMouseMove: callEventHandlers(ownProps?.onMouseMove, hoverTriggerHandler.onMouseMove),
./popover/PopoverTrigger.js:          onMouseLeave: callEventHandlers(ownProps?.onMouseLeave, hoverTriggerHandler.onMouseLeave),
./popover/PopoverContent.js:      onBlur: callEventHandlers(onBlurProp, onBlur),
./popover/PopoverContent.js:      onMouseEnter: callEventHandlers(onMouseEnter, () => {
./popover/PopoverContent.js:      onMouseLeave: callEventHandlers(onMouseLeave, () => {
./popover/PopoverContent.js:    onKeyDown: callEventHandlers(onKeyDown, event => {
./accordion/AccordionToggle.js:      onClick={callEventHandlers(onClickProp, context?.onToggle)}
./tooltip/TooltipTrigger.js:        onBlur: callEventHandlers(ownProps?.onBlur, handleBlur),
./tooltip/TooltipTrigger.js:        onClick: callEventHandlers(ownProps?.onClick, handleClick),
./tooltip/TooltipTrigger.js:        onFocus: callEventHandlers(ownProps?.onFocus, handleFocus),
./tooltip/TooltipTrigger.js:        onMouseDown: callEventHandlers(ownProps?.onMouseDown, handleMouseDown),
./tooltip/TooltipTrigger.js:        onMouseEnter: callEventHandlers(ownProps?.onMouseEnter, handleMouseEnter),
./tooltip/TooltipTrigger.js:        onMouseLeave: callEventHandlers(ownProps?.onMouseLeave, handleMouseLeave),
```

**Comment by trend-derek-hawker** [comment]
Certainly. There's no time rush for this and we found a simple workaround that conditionally adds the onClick callback when !disabled. 

I created an issue for this as I thought it was more ergonomic if the default behaviour didn't trigger the onClick.

**Comment by cheton** [comment]
@mt-derekh

Thank you for raising this issue. In fact, the internal event handlers will stop processing if either `event.defaultPrevented` is set to true or the component received the `disabled` prop. See below:

---

**callEventHandlers**
```js
export const callEventHandlers = (...fns) => {
  return function mergedFn(event) {
    fns.some((fn) => {
      fn?.(event);
      return event?.defaultPrevented;
    });
  };
};
```

**MenuItem**
```js
<Box
  ref={ref}
  role={role}
  tabIndex={-1}
  disabled={disabled}
  aria-disabled={disabled}
  onClick={callEventHandlers(onClick, event => {
    if (disabled) {
      event.stopPropagation();
      event.preventDefault();
      return;
    }
    if (closeOnSelect) {
      ensureFunction(closeMenu)();
    }
  })}
  onKeyDown={callEventHandlers(onKeyDown, event => {
    if (disabled) {
      return;
    }
    if (event.key === 'Enter' || event.key === ' ') {
      event.preventDefault();

      if (onClick) {
        ensureFunction(onClick)();
      }

      if (closeOnSelect) {
        ensureFunction(closeMenu)();
      }
    }
  })}
  {...styleProps}
  {...rest}
/>
```

---

The only uncertainty is that we don't actively stop caller-provided event handler for the disabled state, but leaving the decision to the caller. Currently this behavior opens a way for developers to do something no matter it's disabled or not, although I can't find a good example. 😄 

This might cause a breaking change to some fundamental components. I will bring the discussion to some FED team members to come out a decision.

**Comment by cheton** [comment]
We decided to use `pointer-events: none` to make disabled components ignore pointer events. But we need to confirm with visual designers to check whether the cursor should appear as `not-allowed` for the disabled state.

https://user-images.githubusercontent.com/447801/184461080-92138880-0c20-4376-b36a-a8c1bfb6c2f4.mp4

**Comment by cheton** [comment]
The `callEventHandlers` function is provided in `@tonic-ui/utils`, it's used to merge multiple event handlers and ensure that they are executed in sequence, while stopping execution if any of the handlers call `event.preventDefault()`.

```js
const callEventHandlers = (...fns) => {
  return function mergedFn(event) {
    fns.some((fn) => {
      fn?.(event);
      return event?.defaultPrevented;
    });
  };
};
```

The updated `MenuItem` component will include a `preventDefaultIfDisabled` handler function that prevents the default event behavior when the `disabled` prop is set to true. 

```jsx
import { ariaAttr, callEventHandlers } from '@tonic-ui/utils';
import { ensureFunction } from 'ensure-type';
import React, { forwardRef } from 'react';
import { Box } from '../box';
import { useMenuItemStyle } from './styles';
import useMenu from './useMenu';

const MenuItem = forwardRef((
  {
    disabled,
    onClick,
    onKeyDown,
    role = 'menuitem',
    ...rest
  },
  ref,
) => {
  const menuContext = useMenu(); // context might be an undefined value
  const {
    closeOnSelect,
    onClose: closeMenu,
  } = { ...menuContext };
  const styleProps = useMenuItemStyle();
  const preventDefaultIfDisabled = (event) => {
    if (disabled) {
      event.preventDefault();
    }
  };

  return (
    <Box
      ref={ref}
      role={role}
      tabIndex={-1}
      disabled={disabled}
      aria-disabled={ariaAttr(disabled)}
      onClick={callEventHandlers(preventDefaultIfDisabled, onClick, event => {
        if (closeOnSelect) {
          ensureFunction(closeMenu)();
        }
      })}
      onKeyDown={callEventHandlers(preventDefaultIfDisabled, onKeyDown, event => {
        if (event.key === 'Enter' || event.key === ' ') {
          event.preventDefault();

          if (onClick) {
            ensureFunction(onClick)();
          }

          if (closeOnSelect) {
            ensureFunction(closeMenu)();
          }
        }
      })}
      {...styleProps}
      {...rest}
    />
  );
});
```

**Comment by cheton** [comment]
@mt-derekh 
This issue will be resolved in PR https://github.com/trendmicro-frontend/tonic-ui/pull/698

**Comment by cheton** [comment]
Refrain from using `preventDefaultIfDisabled`, as mentioned in https://github.com/trendmicro-frontend/tonic-ui/issues/613#issuecomment-1429285282, to reduce unnecessary complexity. It's worth noting that `MenuItem`, `MenuToggle`, `Tab` have been restructured with the composition of `ButtonBase` in #698.
